src/pip/_internal/resolution/resolvelib/resolver.py

```diff
@@ -147,6 +152,29 @@ def resolve(self, root_reqs, check_supported_wheels):
                 ireq.should_reinstall = True
             elif dist_is_editable(installed_dist) != candidate.is_editable:
                 ireq.should_reinstall = True
+            elif candidate.source_link.is_file:
+                if candidate.source_link.is_wheel:
+                    logger.info(
+                        "%s is already installed with the same version as the "
+                        "provided wheel. Use --force-reinstall to force an "
+                        "installation of the wheel.",
+                        ireq.name,
+                    )
+                    continue
+                if is_archive_file(candidate.source_link.file_path):
+                    reason = (
+                        "Source distribution is being reinstalled despite an "
+                        "installed package having the same name and version as "
+                        "the installed package."
+                    )
```

Comment:
`is_archive_file` isn't sufficient to distinguish between sdists and archived source trees (e.g., zip files from a github repo). We may need to be prepared to split out non-sdist .tar.gz files if we get pushback over this. (I was only arguing for deprecating the reinstall of sdists).